java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java

```diff
@@ -72,15 +73,53 @@ public void visitNode(Tree tree) {
     if (methodTree.block() != null && "equals".equals(methodTree.symbol().name()) && methodTree.parameters().size() == 1) {
       Symbol parameterSymbol = methodTree.parameters().get(0).symbol();
       if (parameterSymbol.type().is("java.lang.Object")) {
-        ExpressionVisitor visitor = new ExpressionVisitor(parameterSymbol);
-        methodTree.accept(visitor);
-        if (!visitor.typeChecked) {
-          addIssue(tree, "Add a type test to this method.");
+        CastVisitor castVisitor = new CastVisitor(parameterSymbol);
+        methodTree.accept(castVisitor);
+        if (castVisitor.hasCast) {
+          ExpressionVisitor expressionVisitor = new ExpressionVisitor(parameterSymbol);
+          methodTree.accept(expressionVisitor);
+          if (!expressionVisitor.typeChecked) {
+            addIssue(tree, "Add a type test to this method.");
+          }
         }
       }
     }
   }
 
+  private static ExpressionTree removeParenthesis(ExpressionTree tree) {
+    ExpressionTree result = tree;
+    while (true) {
+      if (result.is(Tree.Kind.PARENTHESIZED_EXPRESSION)) {
+        result = ((ParenthesizedTree) result).expression();
+      } else {
+        return result;
+      }
+    }
+  }
+
+  private static class CastVisitor extends BaseTreeVisitor {
+    private final Symbol parameterSymbol;
+    boolean hasCast;
+
+    public CastVisitor(Symbol parameterSymbol) {
+      this.parameterSymbol = parameterSymbol;
+    }
+
+    @Override
+    public void visitTypeCast(TypeCastTree tree) {
+      if (isArgument(tree.expression())) {
+        hasCast = true;
+      } else {
+        super.visitTypeCast(tree);
+      }
+    }
+
+    private boolean isArgument(ExpressionTree tree) {
```

Comment:
This method is duplicated, maybe create a static version with an expression tree and the symbol as parameter.